File: frontend/src/lib/public-settings.ts
import api from "@/lib/api";

export type PublicSettings = {
  project_name: string;
  activity_title: string;
  activity_desc: string;
  default_redirect_url: string;
  customer_service_enabled: boolean;
  customer_service_whatsapp: string;
  customer_service_telegram: string;
};

const DEFAULTS: PublicSettings = {
  project_name: "GroundRewards",
  activity_title: "Lucky Wheel",
  activity_desc: "",
  default_redirect_url: "",
  customer_service_enabled: false,
  customer_service_whatsapp: "",
  customer_service_telegram: "",
};

let cache: Promise<PublicSettings> | null = null;

export async function getPublicSettings(): Promise<PublicSettings> {
  if (!cache) {
    cache = api
      .get<PublicSettings>("/api/public/settings")
      .then((r) => ({ ...DEFAULTS, ...r.data }))
      .catch(() => DEFAULTS);
  }
  return cache;
}

export function clearPublicSettingsCache() {
  cache = null;
}

File: backend/app/routers/public_settings.py
from fastapi import APIRouter, Depends
from motor.motor_asyncio import AsyncIOMotorDatabase

from app.database import get_db

router = APIRouter()

_WHITELIST = {
    "project_name": "GroundRewards",
    "activity_title": "Lucky Wheel",
    "activity_desc": "",
    "default_redirect_url": "",
    "customer_service_enabled": False,
    "customer_service_whatsapp": "",
    "customer_service_telegram": "",
}


@router.get("/settings")
async def get_public_settings(db: AsyncIOMotorDatabase = Depends(get_db)):
    docs = db.system_settings.find({"key": {"$in": list(_WHITELIST.keys())}})
    found: dict = {}
    async for doc in docs:
        found[doc["key"]] = doc.get("value")
    return {key: found.get(key, default) for key, default in _WHITELIST.items()}
